libuuu/buffer.cpp

```diff
@@ -767,14 +767,84 @@ int bz_async_load(string filename, shared_ptr<FileBuffer> p)
 	return 0;
 }
 
-int FSBz2::load(string backfile, string filename, shared_ptr<FileBuffer>p, bool async)
+bool is_pbzip2_file(string filename)
 {
-	if (filename != "*")
+	shared_ptr<FileBuffer> file=get_file_buffer(filename);
+	uint64_t filesize= file->size();
+	uint64_t readsize= (filesize< (1024*1024) )? filesize:(1024*1024); //read at most 1MB, because maximum block size is 900kb
+	int header_num=0;
+	uint8_t* ptr= file->data();
+	for(size_t i =0 ; i < readsize ; i++)
 	{
-		set_last_err_string("bz just support . decompress itself");
+		if(ptr[0]=='B'&& ptr[1]=='Z'&& ptr[2]=='h' && ptr[4]=='1'&& ptr[5]=='A' && ptr[6]=='Y' && ptr[7]=='&' && ptr[8]=='S'&& ptr[9]=='Y')
+		{
+			header_num++;
+		}
+		ptr++;
+	}
+	if(header_num>1)
+		return true;
+	else
+		return false;
+}
+
+int decompress_single_thread(string name,shared_ptr<FileBuffer>p)
+{
+	uint8_t* decompressed_file;
+	uint64_t decompressed_size;
+
+	uint8_t* file;
+	uint64_t size;
+
+	shared_ptr<FileBuffer> filebuffer=get_file_buffer(name);
+
+	size=filebuffer->size();
+	file=filebuffer->data();
+
+	decompressed_file=p->data();
+	decompressed_size=0;
+	p->reserve(7*size);//the usual compressed ratio is about 18%, so 7*18% > 100%
+
+	bz_stream* strm= (bz_stream*)malloc(sizeof(bz_stream));
+	strm->bzalloc  = NULL;
+    strm->bzfree   = NULL;
+	strm->opaque   = NULL;
+
+	int ret;
+	ret = BZ2_bzDecompressInit (strm,0, 0 );
+	if (ret != BZ_OK)
 		return -1;
+	strm->next_in  = (char*)file;
```

Comment:
filebuffer

Reply:
i realized already used filebuffer for the FileBuffer ptr i created, so i rename "file" to "compressed_file", "size" to "compressed_size" for easier understanding.